### PR TITLE
erlcloud_sqs no longer calls erlang:error()

### PR DESCRIPTION
**Pull request by lilrooness**
### Problem
- erlcloud_sqs uses erlang:error() - https://github.com/alertlogic/erlcloud/issues/46
### Solution
- Errors are now returned from API functions as `{error, Reason}`

### CODE REVIEW

File: src/erlcloud_sqs.erl [hunk 1]
```diff
@@ -65,13 +65,18 @@ configure(AccessKeyID, SecretAccessKey, Host) ->
...
 
--spec add_permission/4 :: (string(), string(), sqs_acl(), aws_config()) -> ok.
+-spec add_permission/4 :: (string(), string(), sqs_acl(), aws_config()) -> ok | {error, term()}.
 add_permission(QueueName, Label, Permissions, Config)
   when is_list(QueueName),
        is_list(Label), length(Label) =< 80,
        is_list(Permissions) ->
-    sqs_simple_request(Config, QueueName, "AddPermission",
-                       [{"Label", Label}|erlcloud_aws:param_list(encode_permissions(Permissions), {"AWSAccountId", "ActionName"})]).
+    case sqs_simple_request(Config, QueueName, "AddPermission",
```
Comment (kkuzmin) [inline src/erlcloud_sqs.erl:73]: There is the same case statement inside `sqs_simple_request()`. What is the point of duplicating it?

File: src/erlcloud_sqs.erl [hunk 2]
```diff
@@ -367,19 +388,28 @@ send_message(QueueName, MessageBody, DelaySeconds, Config)
...
-    sqs_request(Config, QueueName, Action, Params),
-    ok.
+    case sqs_request(Config, QueueName, Action, Params) of
+        {error, Reason} ->
+            {error, Reason};
+        {ok, _Body} ->
+            ok
+    end.
 
 sqs_xml_request(Config, QueueName, Action, Params) ->
```
Comment (kkuzmin) [inline src/erlcloud_sqs.erl:414]: What about `erlang:error()` inside this function?

**Comment by nalundgaard** [comment]
I'm not sure this is a safe thing. It looks very broken to me. #46 called for changing to `throw`, not changing error returns to tuples. Although, IMO, that is also a bad idea given the situation.

This is a major breaking change for `erlcloud_sqs` clients who either:
- don't explicitly check the ok return value of `erlcloud_sqs` calls (i.e., they don't write `ok = erlcloud_sqs:set_queue_attributes(QueueName, Attributes)`). This is reasonable client behavior, because the code as it is now will crash with an `error` exception in the event of a problem. And this change will break this behavior by making it silently fail, very likely without even logging anything.
- explicitly catch the `error` exception, e.g.
  
  ``` erlang
  try
      erlcloud_sqs:set_queue_attributes(QueueName, Attributes)
  catch
      error:{aws_error, _Reason} = AwsError->
         handle_sqs_error(AwsError)
  end
  ```
  
  These poor souls will get hard crashes instead of seeing their error handling code executed.

changing this behavior because you don't like it, in a repo that we share publicly, is a misguided idea that will diverge our code base from the public repo, which has responsibilties to a wide array of consumers. If we can't stand this interface, I suggest we write erlcloud_sqs2. Or at least make this change against the public repo and get wider agreement on it.

**Comment by lilrooness** [comment]
Point taken, closing

**Pull request closed without merging**